draft-ietf-jsonpath-base.md

```diff
@@ -911,8 +943,129 @@ size doesn't
+
+
+### Filter Selector
+
+#### Syntax
+
+The filter selector has the shape `'[?<expr]'`. It works via iterating over container values, i.e. arrays and objects. 
+
+~~~~ abnf
+filter-selector    = "[?" boolean-expr "]"
+~~~~
+
+During iteration process each array element or object member is visited and its value &ndash; accessible via symbol `'@'` &ndash; or one of its descendants &ndash; uniquely defined by a relative path &ndash; is tested against a boolean expression `boolean-expr`.
+
+If the result is `true`, the current item, represented by `'@'`, is selected. In case of a `false` result, it is not.
+
+~~~~ abnf
+boolean-expr = *logical-expr
+logical-expr = [neg-op] ["("] comp-expr *[logical-op comp-expr] [")"]
+neg-op       = "!"                                  ; not operator
+logical-op   = "||" / "&&"                          ; logical operator
+comp-expr    = (rel-path-val /
+                calc-val) [(comp-op comparable /  ; comparison
+                            regex-op regex     /  ; RegEx test
+                            in-op container )]    ; containment test
+comp-op      = "==" / "!=" /                        ; comparison ...
+               "<"  / ">"  /                        ; operators
+               "<=" / ">="
+regex-op     = "~="                                 ; RegEx match
+in-op        = " in "                               ; in operator
+comparable   = number / quoted-string /             ; primitive ...
+               true / false / null /                ; values only
+               rel-path-val /                       ; descendant value
+               calc_val /                           ; calculated value
+               json-path                            ; any value
+               
+rel-path-val = "@" *(dot-selector / index-selector) 
+calc_val     = func "(" [rel-path-val / json-path] ")"
+func         = "index"
+~~~~
+
+Notes:
+
+* Parenthesis can be used with `boolean-expr` for grouping. So filter selection syntax in the original proposal `'[?(<expr)]'` is naturally contained in the current lean syntax `'[?<expr]'` as a special case.
+* Comparisons are restricted to primitive values `number`, `string`, `true`, `false`, `null`. Comparisons with complex values will fail, i.e. no selection occurs.
+* Implicite type conversions during comparisons are not performed. So `"13 == '13'"` selects nothing.
+* A member or element value by itself is *falsy* only, if it does not exist. Otherwise it is *truthy*, resulting in its value. To be more specific explicite comparisons are necessary. This existence test &ndash; as an exception of the general rule &ndash; also works with complex values.
+* Regular expression tests can be applied to `string` values only.
+* Containment tests work with arrays and objects.
+* Explicite boolean type conversion is done by the not operator `neg-op`.
+* The behaviour of operators is consistent with the 'C'-family of programming languages.
```

Comment:
This statement seems out of place in the syntax section. Do we need an "Informal Semantics" section?

Reply:
will be moved. Not sure when to move to semantic or informal semantics section.

Reply:
@goessner  wrote:
> will be moved. Not sure when to move to semantic or informal semantics section.

 If by "when" you meant "whether", I think the statement should be moved to a new informal semantics section.

Reply:
It should perhaps read better:
```
Syntax and behaviour of operators is consistent with the 'C'-family of programming languages.
```
Otherwise we might need to add a table of operator precedence ... which would be no problem though.

Reply:
FYI: Have a look at how RFC 8949 solves that problem:
https://www.rfc-editor.org/rfc/rfc8949.html#section-1.2-6

The IESG will insist on a reference to the C standard, I can tell you this much :-)

Reply:
@Glyn: By "when" I mean "whenever" in comparable situations, should it goto

* syntax
* semantics
* informal semantics

Are there specific rules especially to allocate to the last two.

Reply:
> @glyn: By "when" I mean "whenever" in comparable situations, should it goto
> 
>     * syntax
> 
>     * semantics
> 
>     * informal semantics
> 
> 
> Are there specific rules especially to allocate to the last two.

No rules, but the idea with array slices was to introduce the semantics gently in the (informative) informal semantics section and then spell out the (normative) semantics, possibly more indigestibly, in the semantics section. But I am not wedded to this convention so long as we don't confuse normative and informative text.